src/loader/utils/load_batch.php

```diff
@@ -28,10 +28,24 @@ function load_batch(array $config, array $transformedevents, callable $loader) {
         $loadedevents = construct_loaded_events($transformedevents, true);
         return $loadedevents;
     } catch (\Exception $e) {
+        $batchsize = count($transformedevents);
         $logerror = $config['log_error'];
-        $logerror("Failed load for event id #" . $eventobj->id . ": " .  $e->getMessage());
+        $logerror("Failed load batch (" . $batchsize . " events)" .  $e->getMessage());
         $logerror($e->getTraceAsString());
-        $loadedevents = construct_loaded_events($transformedevents, false);
+
+        // In the event of a 400 error, recursively retry sending statements in increasingly
+        // smaller batches so that only the actual bad data fails.
+        if ($batchsize === 1 || $e->getCode() !== 400 || $config['lrs_resend_failed_batches'] !== '1') {
+            $loadedevents = construct_loaded_events($transformedevents, false);
+        } else {
+            $newconfig = $config;
+            $newconfig['lrs_max_batch_size'] = round($batchsize / 2);
```

Comment:
Really like the way you solved this recursively with a modified copy of `$config`.